src/bp/nlu-core/engine.ts

```diff
@@ -1,48 +1,63 @@
 import { MLToolkit, NLU } from 'botpress/sdk'
 import crypto from 'crypto'
 import _ from 'lodash'
+import LRUCache from 'lru-cache'
+import sizeof from 'object-sizeof'
 
 import { EntityCacheManager } from './entities/entity-cache-manager'
 import { initializeTools } from './initialize-tools'
 import DetectLanguage from './language/language-identifier'
-import { deserializeModel, PredictableModel, serializeModel } from './model-manager'
+import { deserializeModel, PredictableModel, serializeModel } from './model-serializer'
 import { Predict, PredictInput, Predictors, PredictOutput } from './predict-pipeline'
 import SlotTagger from './slots/slot-tagger'
 import { isPatternValid } from './tools/patterns-utils'
 import { computeKmeans, ProcessIntents, TrainInput, TrainOutput } from './training-pipeline'
-import { TrainingCanceledError, TrainingWorkerQueue } from './training-worker-queue'
+import { TrainingWorkerQueue } from './training-worker-queue'
 import { EntityCacheDump, Intent, ListEntity, PatternEntity, Tools } from './typings'
 
 const trainDebug = DEBUG('nlu').sub('training')
 
-export default class Engine implements NLU.Engine {
-  private static _tools: Tools
-  private static _trainingWorkerQueue: TrainingWorkerQueue
+interface LoadedModel {
+  model: PredictableModel
+  predictors: Predictors
+  entityCache: EntityCacheManager
+}
+
+const DEFAULT_OPTIONS: Options = {
+  maxCacheSize: 250000000 // 250mb of model cache
+}
 
-  private predictorsByLang: _.Dictionary<Predictors> = {}
-  private modelsByLang: _.Dictionary<PredictableModel> = {}
-  private entitiesCacheByLang: _.Dictionary<EntityCacheManager> = {}
+interface Options {
+  maxCacheSize: number
+}
+
+export default class Engine implements NLU.Engine {
+  private _tools!: Tools
+  private _trainingWorkerQueue!: TrainingWorkerQueue
 
-  constructor(private botId: string, private logger: NLU.Logger) {}
+  private modelsById: LRUCache<string, LoadedModel>
 
-  // NOTE: removed private in order to prevent important refactor (which will be done later)
-  public static get tools() {
-    return this._tools
+  constructor(opt?: Partial<Options>) {
+    const options: Options = { ...DEFAULT_OPTIONS, ...opt }
+    this.modelsById = new LRUCache({
+      max: options.maxCacheSize,
+      length: sizeof // ignores size of functions, but let's assume it's small
```

Comment:
sizeof ([object-sizeof](https://www.npmjs.com/package/object-sizeof)) estimates the size of an object based on amount of numbers and characters. It can't compute the size of unserializable fields like functions, but for the rest it works fines. 

I double checked with a breakpoint just to make sure and the estimation seems fine; for the bpds-intent-A model its about ~18mb.